packages/protocol/test-sol/governance/validators/Validators.t.sol

```diff
@@ -0,0 +1,3204 @@
+
+    ph.mockSuccess(ph.PROOF_OF_POSSESSION(), abi.encodePacked(_validator, blsPublicKey, blsPop));
+
+    vm.prank(_validator);
+    validators.registerValidator(_ecdsaPubKey, blsPublicKey, blsPop);
+    validatorRegistrationEpochNumber = validators.getEpochNumber();
+    return _ecdsaPubKey;
+  }
+
+  function _registerValidatorGroupHelper(address _group, uint256 numMembers) internal {
+    if (!accounts.isAccount(_group)) {
+      vm.prank(_group);
+      accounts.createAccount();
+    }
+
+    lockedGold.setAccountTotalLockedGold(
+      _group,
+      originalGroupLockedGoldRequirements.value.mul(numMembers)
+    );
+
+    vm.prank(_group);
+    validators.registerValidatorGroup(commission.unwrap());
+  }
+
+  function _registerValidatorGroupWithMembers(address _group, uint256 _numMembers) public {
+    _registerValidatorGroupHelper(_group, _numMembers);
+
+    for (uint256 i = 0; i < _numMembers; i++) {
+      if (i == 0) {
+        _registerValidatorHelper(validator, validatorPk);
+
+        vm.prank(validator);
+        validators.affiliate(group);
+
+        vm.prank(group);
+        validators.addFirstMember(validator, address(0), address(0));
+      } else {
+        uint256 _validator1Pk = i;
+        address _validator1 = vm.addr(_validator1Pk);
+
+        vm.prank(_validator1);
+        accounts.createAccount();
+        _registerValidatorHelper(_validator1, _validator1Pk);
+        vm.prank(_validator1);
+        validators.affiliate(group);
+
+        vm.prank(group);
+        validators.addMember(_validator1);
+      }
+    }
+  }
+
+  function _max1(uint256 num) internal pure returns (FixidityLib.Fraction memory) {
+    return num > FixidityLib.fixed1().unwrap() ? FixidityLib.fixed1() : FixidityLib.wrap(num);
+  }
+
+  function _safeExponent(FixidityLib.Fraction memory base, FixidityLib.Fraction memory exponent)
+    internal
+    pure
+    returns (uint256)
+  {
+    if (FixidityLib.equals(base, FixidityLib.newFixed(0))) return 0;
+    if (FixidityLib.equals(exponent, FixidityLib.newFixed(0))) return FixidityLib.fixed1().unwrap();
+
+    FixidityLib.Fraction memory result = FixidityLib.fixed1();
+
+    for (uint256 i = 0; i < exponent.unwrap(); i++) {
+      if (FixidityLib.multiply(result, base).value < 1) revert("SafeExponent: Overflow");
+
+      result = FixidityLib.multiply(result, base);
+    }
+    return result.unwrap();
+  }
+
+  function _calculateScore(uint256 _uptime, uint256 _gracePeriod) internal view returns (uint256) {
+    return
+      _safeExponent(
+        _max1(_uptime.add(_gracePeriod)),
+        FixidityLib.wrap(originalValidatorScoreParameters.exponent)
+      );
+  }
+}
+
+contract ValidatorsTest_Initialize is ValidatorsTest {
+  function test_ShouldhaveSetTheOwner() public {
+    assertEq(validators.owner(), owner, "Incorrect Owner.");
+  }
+
+  function test_Reverts_WhenCalledMoreThanOnce() public {
+    vm.expectRevert();
```

Comment:
Can we expect the error message ?

Reply:
This was left empty, because the the received error in this test is overridden by the tunnel.

require(success, "unsuccessful tunnel call");

I felt that adding that revert message was a bit misleading. I can change it to expect `unsuccessful tunnel call` if you'd prefer.